### PR TITLE
Refactor of strategies following two-step framework of scoring and sampling

### PR DESCRIPTION
**Pull request by thomasgaudelet**
## What is the goal of this PR?

This PR introduces changes to separate scoring and sampling responsibilities in strategies following a two-step framework design that most active learning strategies abide by.

## What are the changes implemented in this PR?

- BatchModeSampler class and child class that provide the abstraction for batch active learning sampling. We support two at the moment: TopKSampler, a greedy sampling algorithm based on scores, and ProbabilisticSampler that treats the scores as probability distribution over the indices.
- Informativeness Scorer classes that take on the responsibility of scoring samples from model outputs

The strategies are refactored to adapt to the new design choice.

### CODE REVIEW

File: pyrelational/informativeness/decorators.py
Comment (paulmorio) [inline pyrelational/informativeness/decorators.py:1]: Looks good.

General question: would it be useful to be explicit about the parameters in `torch.allclose` i.e.  rtol=1e-05, atol=1e-08, equal_nan=False?

File: pyrelational/samplers/abstract_sampler.py
Comment (paulmorio) [inline pyrelational/samplers/abstract_sampler.py:1]: This is great, just a thought on naming.

Whilst sampling is correct, I think it would make sense to lengthen the name to `*BatchSampler`. My rationale is that it would help avoid ambiguity with other samplers (lower mental load) and reinforces whats already in AL literature.

File: pyrelational/samplers/abstract_sampler.py [hunk 1]
```diff
@@ -0,0 +1,43 @@
...
+    def __call__(self, scores: Tensor, indices: List[int], num_samples: int) -> List[int]:
+        """Sample a subset of indices based on the scores.
+
+        This method should be implemented by the subclasses.
+        :return: List of sampled indices.
+        """
+        pass
+
+
+class DeterministicSampler(AbstractSampler):
```
Comment (paulmorio) [inline pyrelational/samplers/abstract_sampler.py:23]: Again just a discussion on naming. I think it would be good to just call it a `TopKSampler`. There are other deterministic batch samplers out there.

File: pyrelational/samplers/abstract_sampler.py [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+
+class ProbabilisticSampler(AbstractSampler):
+    """Probabilistic sampler based on the scores."""
+
+    def __call__(self, scores: Tensor, indices: List[int], num_samples: int) -> List[int]:
+        """Sample a subset of indices deriving a distribution from the scores.
+
+        :return: List of sampled indices.
+        """
+        return [indices[i] for i in torch.multinomial(scores, num_samples, replacement=False).tolist()]
```
Comment (paulmorio) [inline pyrelational/samplers/abstract_sampler.py:43]: With `replace=False` it will give an error when there are less items in `scores` than `num_samples`. Our current code base will always select the amount available when this happens, so maybe a try-catch pattern with the catch-pattern adjusting the num_samples (or just returning all of the samples).

File: pyrelational/samplers/abstract_sampler.py
Comment (thomasgaudelet) [reply pyrelational/samplers/abstract_sampler.py:1]: hmm, I don't like using the word batch as I think this is also misleading. Maybe QuerySampler?

File: pyrelational/samplers/abstract_sampler.py (hunk 1 above)
Comment (thomasgaudelet) [reply pyrelational/samplers/abstract_sampler.py:23]: This would kind of imply there would be a K hyperparameter that is fixed

Comment (paulmorio) [reply pyrelational/samplers/abstract_sampler.py]: The `num_samples` would be this K. I think the naming would be ok in the same way it is used in https://pytorch-geometric.readthedocs.io/en/latest/generated/torch_geometric.nn.pool.TopKPooling.html

File: pyrelational/informativeness/abstract_scorers.py [hunk 3]
```diff
@@ -0,0 +1,78 @@
...
+
+
+class DecoratedClassificationScorerMeta(type):
+    """Metaclass for classification scorers."""
+
+    def __new__(cls, name: str, bases: tuple[Any, ...], dct: dict[str, Any]) -> "DecoratedClassificationScorerMeta":
+        """Decorate the `__call__` method with the `require_2d_tensor` and `require_probabilities` decorators."""
+        for key, value in dct.items():
+            if callable(value) and key == "__call__":
+                dct[key] = require_probabilities(require_2d_tensor(value))
```
Comment (paulmorio) [inline pyrelational/informativeness/abstract_scorers.py:32]: Sorry I think I need a bit of hand-holding to get through this part. Why do we need informativeness scores to be 2D probability distributions? What is the role of the dictionary?

**Review by paulmorio — COMMENTED** [review]
Thanks for doing this on my really vague idea! It's a great realisation and makes me think a little about how to go forward with it. Left some comments or points for discussion.

File: pyrelational/samplers/abstract_sampler.py (hunk 1 above)
Comment (thomasgaudelet) [reply pyrelational/samplers/abstract_sampler.py]: it's not fixed though, as we pass it to __call__. I think Deterministic makes more sense if we call the other Probabilistic

File: pyrelational/informativeness/abstract_scorers.py (hunk 3 above)
Comment (thomasgaudelet) [reply pyrelational/informativeness/abstract_scorers.py]: So 2D requirements is actually breaking the repo, so I removed it. But something we should think about though. In classification, it looks like most scoring methods expects batch by number of classes matrix input, the third dimensions which would be for instance number of estimators is not clearly used. We also have that .mean(0) always in ClassificationStrategy post train_and_infer call that collapses the estimator dimension in all cases (that probably should be removed tbh, putting the responsibility on the user to do that rather that making such an assumptions ourselves, I keep it here because it's not the object of this PR but something to keep in mind).

The Metaclass is used here to automatically decorate the __call__ methods of any child classes instead of having to redecorate the method each time. Maybe something to relax if there's a score that doesn't require probabilities.

File: pyrelational/informativeness/regression_scorers.py [hunk 4]
```diff
@@ -0,0 +1,159 @@
...
+        :param axis: index of the axis along which the repeats are
+        :return:  pytorch tensor of scores
+        """
+        if mean is None:
+            assert x is not None, "both x and mean are None, cannot compute."
+            return self.compute_mean(x)
+        return mean
+
+
+class RegressionLeastConfidence(AbstractRegressionScorer):
```
[the remark this answers is not in the document]
Comment (thomasgaudelet) [reply pyrelational/informativeness/regression_scorers.py]: We have both LeastConfidence and MarginConfidence in classification score.

I changed to StandardDeviation for regression, we need a better name for the Average/MeanPrediction one though.

File: pyrelational/strategies/strategy.py [hunk 5]
```diff
@@ -0,0 +1,46 @@
...
+        :param data_manager: A pyrelational data manager
+            which keeps track of what has been labelled and creates data loaders for
+            active learning
+        :param model_manager: A pyrelational model manager
+            which wraps a user defined ML model to handle instantiation, training, testing,
+            as well as uncertainty quantification
+
+        :return: list of indices to annotate
+        """
+        output = self.train_and_infer(data_manager=data_manager, model_manager=model_manager).mean(0)
```
[the remark this answers is not in the document]
Comment (thomasgaudelet) [reply pyrelational/strategies/strategy.py]: outdated

File: pyrelational/samplers/abstract_sampler.py
Comment (paulmorio) [reply pyrelational/samplers/abstract_sampler.py:1]: Query learning, and Query sampling are sometimes synonymous with Active learning (Settles textbook). "Batch-mode" is the key term I find most associated with research presenting new batch construction algorithms.

Comment (paulmorio) [reply pyrelational/samplers/abstract_sampler.py:1]: QueryBatchSampler? 😵‍💫

**Review by paulmorio — DISMISSED** [review]
Great! Thanks for all the changes, really like the approach to the new strategies and class definitions. Definitely aligns with the new directions. I agree it's difficult with the representative sampling. There's probably some way or perspective on the representative sampling which would fit under a scoring/sampling metric, but it would be non-practical which is the point of the library.

**Review by paulmorio — APPROVED** [review]
Looks good to me!

**Review by paulmorio — APPROVED** [review]
Approving merge
